source/extensions/clusters/logical_dns/logical_dns_cluster.cc

```diff
@@ -152,7 +152,13 @@ void LogicalDnsCluster::startResolve() {
             final_refresh_rate = addrinfo.ttl_;
           }
           if (dns_jitter_ms_.count() != 0) {
-            final_refresh_rate += std::chrono::milliseconds(random_.random()) % dns_jitter_ms_;
+            // We need to wrap with std::chrono::abs because
+            // random() returns an uint64 that is interpreted as a long by
+            // std::chrono::milliseconds. If the random() return value has a leading 1 in its binary
+            // representation, then it will be interpreted as a negative long, potentially causing
+            // final_refresh_rate to be negative. This causes Envoy to crash.
+            final_refresh_rate +=
+                std::chrono::abs(std::chrono::milliseconds(random_.random()) % dns_jitter_ms_);
```

Comment:
I think another way to solve this is to use:
`std::chrono::milliseconds(random_.random() % dns_jitter_ms_.count())` dns_jitter_ms is in the range of valid ms so I think the uint % long long will be a smaller positive number in the range of ms.

This _might_ still be possible to overflow in adding the two if the values are very large e.g. `std::numeric_limit<int64>::max()`

Reply:
I think `count()` returns an int64, but I agree.

Reply:
To clarify the latter end of my comment is it still possible for final_refresh_rate to overflow from the addition?

Reply:
yes, BUT only if `dns_jitter_ms` is negative. I will add a validator and test case for this.

Reply:
@KBaichoo 
So I'm fairly confident that `std::chrono::milliseconds(random_.random() % dns_jitter_ms_.count())` wont have any overflow shenanigans. The biggest time value we have have in protobufs is 315,576,000,000, so the most milliseconds that can be represented is 31,557,600,000,000). Then `0 < random_.random() % dns_jitter_ms_.count() < 31,557,600,000,000`.
`std::chrono::milliseconds` is represented internally by an int64, which can take values up to 9,223,372,036,854,775,807. So we should be good.  But,

a) would appreciate a check on my logic
b) kinda want to wrap it all in an `abs` and not think about binary representations/limits.

Reply:
Thank you for thinking through this, this lgtm.